source/server/filter_chain_manager_impl.cc

```diff
@@ -335,27 +343,52 @@ void FilterChainManagerImpl::addFilterChainForServerNames(
 void FilterChainManagerImpl::addFilterChainForApplicationProtocols(
     ApplicationProtocolsMap& application_protocols_map,
     const absl::Span<const std::string* const> application_protocols,
+    const std::vector<std::string>& direct_source_ips,
     const envoy::config::listener::v3::FilterChainMatch::ConnectionSourceType source_type,
     const std::vector<std::string>& source_ips,
     const absl::Span<const Protobuf::uint32> source_ports,
     const Network::FilterChainSharedPtr& filter_chain) {
   if (application_protocols.empty()) {
-    addFilterChainForSourceTypes(application_protocols_map[EMPTY_STRING], source_type, source_ips,
-                                 source_ports, filter_chain);
+    addFilterChainForDirectSourceIPs(application_protocols_map[EMPTY_STRING].first,
+                                     direct_source_ips, source_type, source_ips, source_ports,
+                                     filter_chain);
   } else {
     for (const auto& application_protocol_ptr : application_protocols) {
-      addFilterChainForSourceTypes(application_protocols_map[*application_protocol_ptr],
-                                   source_type, source_ips, source_ports, filter_chain);
+      addFilterChainForDirectSourceIPs(application_protocols_map[*application_protocol_ptr].first,
+                                       direct_source_ips, source_type, source_ips, source_ports,
+                                       filter_chain);
+    }
+  }
+}
+
+void FilterChainManagerImpl::addFilterChainForDirectSourceIPs(
+    DirectSourceIPsMap& direct_source_ips_map, const std::vector<std::string>& direct_source_ips,
+    const envoy::config::listener::v3::FilterChainMatch::ConnectionSourceType source_type,
+    const std::vector<std::string>& source_ips,
+    const absl::Span<const Protobuf::uint32> source_ports,
+    const Network::FilterChainSharedPtr& filter_chain) {
+  if (direct_source_ips.empty()) {
+    addFilterChainForSourceTypes(direct_source_ips_map[EMPTY_STRING], source_type, source_ips,
+                                 source_ports, filter_chain);
+  } else {
+    for (const auto& direct_source_ip : direct_source_ips) {
```

Comment:
Should we (or do we) have API docs somewhere that talk about the O(size) addition here?

Reply:
I'm not sure. @lambdai wdyt?

Reply:
alternately @htuch  / @lizan can weigh in here as shephards

Reply:
Is the question on how do we describe hte memory overheads of building up the nested tries for matching?

If so, I think maybe we should have some discussion on this in the top-level `FilterChainMatch` message comment.

Reply:
I think it may be worth documenting, but probably outside the scope of this PR; this doesn't fundamentally change how it scales, and adds negligible additional memory usage when the new feature isn't being used.